pkg/planner/core/operator/logicalop/logical_datasource.go

```diff
@@ -621,3 +633,90 @@ func preferKeyColumnFromTable(dataSource *DataSource, originColumns []*expressio
+func (ds *DataSource) analyzeFTSFunc() error {
+	if !ds.SCtx().HasFTSFunc() {
+		return nil
+	}
+	idx2FastCheck := make(map[*model.IndexInfo]intset.FastIntSet)
+	idSetForCheck := intset.NewFastIntSet()
+	for _, index := range ds.FtsIndexes {
+		s := intset.NewFastIntSet()
+		for _, col := range index.Columns {
+			s.Insert(int(ds.TableInfo.Columns[col.Offset].ID))
+		}
+		idx2FastCheck[index] = s
+	}
+	var matchedIdx *model.IndexInfo
+	var matchedFunc *expression.ScalarFunction
+	var matchedCondPos int
+	matchedColumns := make([]*expression.Column, 0, 2)
+	for i, cond := range ds.PushedDownConds {
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.FTSMatchWord {
+			if expression.ContainsFullTextSearchFn(cond) {
+				return plannererrors.ErrWrongUsage.FastGen(plannererrors.FTSWrongPlace)
+			}
+			continue
+		}
+		idSetForCheck.Clear()
+		for i := 1; i < len(sf.GetArgs()); i++ {
+			col := sf.GetArgs()[i].(*expression.Column)
+			idSetForCheck.Insert(int(col.ID))
+		}
+		var currentIndex *model.IndexInfo
+		for idx, set := range idx2FastCheck {
+			// The used columns in the FTS function should be a subset of the index columns.
+			if idSetForCheck.SubsetOf(set) {
+				currentIndex = idx
+				break
+			}
+		}
+		// If the index is not found, it means that the FTS function is not valid.
+		if currentIndex == nil {
+			return errors.New("Full text search can only be used with a matching fulltext index and a columnar storage")
+		}
+		// Currently TiDB doesn't support multiple fulltext search functions used with multiple index calls.
+		if matchedIdx != nil {
+			return errors.New("Current TiDB doesn't support multiple fulltext search functions used with multiple index calls")
+		}
+		matchedIdx = currentIndex
+		matchedFunc = sf
+		matchedCondPos = i
+		matchedColumns = matchedColumns[:0]
+		for i := 1; i < len(sf.GetArgs()); i++ {
+			col := sf.GetArgs()[i].(*expression.Column)
+			matchedColumns = append(matchedColumns, col)
+		}
+	}
+	// Fulltext index must be used. So we prune all other possible access paths.
+	ds.PossibleAccessPaths = slices.DeleteFunc(ds.PossibleAccessPaths, func(path *util.AccessPath) bool {
+		return path.StoreType == kv.TiKV
+	})
+	if len(ds.PossibleAccessPaths) == 0 {
+		return plannererrors.ErrWrongUsage.FastGen("Full text search can be only executed in a columnar storage")
+	}
+	// Build protobuf info for the matched index.
+	pbColumns := make([]*tipb.ColumnInfo, 0, len(matchedColumns))
+	for _, col := range matchedColumns {
+		pbColumns = append(pbColumns, tidbutil.ColumnToProto(col.ToInfo(), false, false))
+	}
+	colNames := make([]string, 0, len(matchedColumns))
+	for _, col := range matchedColumns {
+		colNames = append(colNames, col.OrigName)
+	}
+	ds.FtsIndexes[0] = matchedIdx
+	ds.FtsIndexes = ds.FtsIndexes[:1]
+	ds.PushedDownConds = append(ds.PushedDownConds[:matchedCondPos], ds.PushedDownConds[matchedCondPos+1:]...)
```

Comment:
```suggestion
	ds.FtsIndexes = []*model.IndexInfo{matchedIdx}
	ds.PushedDownConds = slices.Delete(ds.PushedDownConds, matchedCondPos, matchedCondPos+1)
```

Reply:
to save memory, the first change will not be accepted this time.